### DISCUSSION TITLE
"Improper" Token

### DISCUSSION THREAD

**Discussion by arn5891** [Rapptz/discord.py#9492] [Q&A] [answered]
I'm using Render to run my bot, and I keep getting an error message that I have the wrong token. The same token seems to work on Replit, and I've already reset the token. <b>I am not using the client secret.</b> My guess is I've declared the token wrong in the secrets tab. The error message seems a bit redundant, but here's the whole thing.
> Traceback (most recent call last):
Jul 10 05:51:35 PM    File "main.py", line 24, in <module>
Jul 10 05:51:35 PM      client.run(str(os.getenv("TOKEN")))
Jul 10 05:51:35 PM    File "/usr/local/lib/python3.8/site-packages/discord/client.py", line 860, in run
Jul 10 05:51:35 PM      asyncio.run(runner())
Jul 10 05:51:35 PM    File "/usr/local/lib/python3.8/asyncio/runners.py", line 44, in run
Jul 10 05:51:35 PM      return loop.run_until_complete(main)
Jul 10 05:51:35 PM    File "/usr/local/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
Jul 10 05:51:35 PM      return future.result()
Jul 10 05:51:35 PM    File "/usr/local/lib/python3.8/site-packages/discord/client.py", line 849, in runner
Jul 10 05:51:35 PM      await self.start(token, reconnect=reconnect)
Jul 10 05:51:35 PM    File "/usr/local/lib/python3.8/site-packages/discord/client.py", line 777, in start
Jul 10 05:51:35 PM      await self.login(token)
Jul 10 05:51:35 PM    File "/usr/local/lib/python3.8/site-packages/discord/client.py", line 612, in login
Jul 10 05:51:35 PM      data = await self.http.static_login(token)
Jul 10 05:51:35 PM    File "/usr/local/lib/python3.8/site-packages/discord/http.py", line 807, in static_login
Jul 10 05:51:35 PM      raise LoginFailure('Improper token has been passed.') from exc
Jul 10 05:51:35 PM  discord.errors.LoginFailure: Improper token has been passed.
Jul 10 05:51:54 PM  [2023-07-10 21:51:54] [INFO    ] discord.client: logging in using static token
Jul 10 05:51:54 PM  Traceback (most recent call last):
Jul 10 05:51:54 PM    File "/usr/local/lib/python3.8/site-packages/discord/http.py", line 803, in static_login
Jul 10 05:51:54 PM      data = await self.request(Route('GET', '/users/@me'))
Jul 10 05:51:54 PM    File "/usr/local/lib/python3.8/site-packages/discord/http.py", line 745, in request
Jul 10 05:51:54 PM      raise HTTPException(response, data)
Jul 10 05:51:54 PM  discord.errors.HTTPException: 401 Unauthorized (error code: 0): 401: Unauthorized

**Comment by No767** [comment, marked as answer]
`os.getenv()` has a chance to return `None` if the environment variable could not be found. This is usually due to not properly loading envs in your environment. Check your Render settings again to ensure that you have it loaded

**Reply by arn5891** [reply]
Thanks, that worked.

**Comment by LostLuma** [comment]
According to their docs secrets are files stored in `/etc/secrets`, so you can not load them via `os.getenv`.

From [Configuring Environment Variables and Secrets](https://render.com/docs/configure-environment-variables#secret-and-environment-configuration-files) on the secrets image:
> You can store secret files (like .env or .npmrc files and private keys) in Render. These files can be accessed during builds and in your code just like regular files.
> All secret files you create are available to read at the root of your repo (or Docker context). They are also available to load by absolute path at /etc/secrets/<filename>.

Do note that this sort of host is not made for hosting Discord bots, so you may run into other issues similarly to the ones people do on replit.
